### ISSUE TITLE
[TW-657] locking=off not effective

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#692]
_Tom Duffy on 2011-03-01T15:44:09Z says:_ 

 First I installed and tested that task was working,  V1.9.3. on FreeBsd 7.3

This freeBsd machine doesn't have a backup mechanism in place (it's a VM that only gets saved every month or so).

Changed home directory to be a NFS share from a RHEL file server that does have daily backup.

Made sure that the user ID and Group ID match between the two systems.

task now no longers works - can't get lock on pending file.

Read bug reports, read that there is a locking=off option.

Added locking=off to .taskrc

retry task ls,  no change, still reports it can't get the lock.

I'm not a NFS guru, but normal file access (read/write) was working fine, and the ownership of the files looks fine.

Is locking=off not working correctly?

In the mean time, I've abandoned the NFS share and will implement a backup by rsyncing the .task dir and rc file to the backed up machine.

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2011-03-01T15:44:09Z
Modified: 2014-02-09T02:06:32Z
```

**Comment by taskwarrior** [comment]
_Cory Donnelly on 2011-03-01T18:33:13Z says:_ 

 This might be a silly question, but are you sure your user account has write access to your task data directory? What options did you pass to mount?

**Comment by taskwarrior** [comment]
_Tom Duffy on 2011-03-01T18:44:18Z says:_ 

 Not a silly question, I asked myself the same thing before suspecting locking=off

The directory is exported as rw from the File Server.

I checked that I can create, edit and delete files under the NFS mounted home directory, and the .task subdirectory, and that all the files I can see there (3 .data files) are chmod 644, and everything with my user:group 
The actual error is:
$ task
Waiting for file lock...
[task list]
Could not lock '/usr/home/tduffy/.task/pending.data'.

**Comment by taskwarrior** [comment]
_Tom Duffy on 2011-03-07T19:05:10Z says:_ 

 I downloaded the tarball for 1.9.4 and complied locally.  git doesn't work behind firewalls, and I couldn't get any of the supposed workarounds to work, gave up on that.

I instrumented in TDB.cpp, method openAndLock(), adding the line:

 std::cout << "mLock is " << mLock << "\n";

just to track what is coming in.

Set locking=off in .taskrc

run lock:

mLock is 1
mLock is 1
mLock is 1
mLock is 0
mLock is 0
mLock is 0
[task list]

<...>

So even with locking set to off, because mLock is set to true on class construction, there seems to be one set of 3 calls to OpenAndLock that is occuring before lock(false) has been called.

**Comment by taskwarrior** [comment]
_Tom Duffy on 2011-03-07T19:58:16Z says:_ 

 The meta-problem here is locking not working over NFS - that turns out to be a configuration issue.
On FreeBSD as a NFS client, it is not sufficient to just enable nfs_client,  the rpc.lockd and rpc.statd daemons must be running too:
in /etc/rc.conf:
1. NFS Client
nfs_client_enable="YES"
rpc_lockd_enable="YES"
rpc_lockd_flags=""
rpc_statd_enable="YES"
rpc_statd_flags=""

Enabling these lets task work correctly across this NFS mount.

But the bug as stated exists, even with locking=off,  task attempts to openAndLock the 3 files at least once with mLock=on

**Comment by taskwarrior** [comment]
_Cory Donnelly on 2011-03-08T14:40:34Z says:_ 

 Thanks for identifying and documenting the FreeBSD NFS config issue as well as pinpointing the cause of the original bug. We'll definitely follow up on this.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-07-09T22:30:54Z says:_ 

 I have looked into the code, but do not have access to an NFS.

The code appears correct regarding locking, but generated a "Could not lock ..." error even when locking was set to off.  This error message has been corrected.  I think this was just a misleading error message.

Tom, if you get a chance, when 2.0 is released I would appreciate it if you would report back with any problems, or better yet, during the beta.  Thanks for all your help.
